added a 'neutron reconstruction' algorithm which creates a neutron candidate that merges all clusters in a given bank

sebouh137 opened the pull request:
### Briefly, what does this PR introduce?
Adds a cluster-merging algorithm that creates a "neutron candidate" (of type ReconstructedParticle) that merges all clusters in a given bank together.  

### What kind of change does this PR introduce?
- [ ] Bug fix (issue #__)
- [X] New feature (issue #1403 )
- [ ] Documentation update
- [ ] Other: __

### Please check if this PR fulfills the following:
- [X] Tests for the changes have been added
- [X] Documentation has been added / updated
- [X] Changes have been communicated to collaborators

### Does this PR introduce breaking changes? What changes might users need to make to their code?
no
### Does this PR change default behavior?
Yes, changes ZDC reconstruction to use ImagingTopo clustering instead of island clustering.   Also tweaked the ImagingTopo algorithm parameters that are used in ZDC recon

wdconinc commented:
@sebouh137 Please fix merge conflicts.

sebouh137 commented:
> @sebouh137 Please fix merge conflicts.

Fixed

ruse-traveler commented on src/algorithms/calorimetry/NeutronReconstructionConfig.h:
```diff
@@ -0,0 +1,12 @@
+
+namespace eicrecon {
+
+struct NeutronReconstructionConfig {
+
+
+};
```
Should there be parameters here? If not, I'd suggest using `WithPodConfig<NoConfig>` and removing this file.

sebouh137 replied:
Done as requested

ruse-traveler reviewed (changes requested):
Thanks for the development, Sebouh! I have a minor fix suggested on the code itself, but I have some larger conceptual suggestions:

- I think this might make more sense alongside the `reco` algorithms since this is taking the final step from detector objects (topoclusters) to physics objects (neutron candidates)
- I would suggest renaming this to something like `FarForwardNeutronReconstrution` since neutrons will also be reconstructed elsewhere in ePIC with very different algorithms.

veprbl commented on src/algorithms/calorimetry/NeutronReconstructionConfig.h:
```diff
@@ -0,0 +1,12 @@
+// Subject to the terms in the LICENSE file found in the top-level directory.
```
```suggestion
// SPDX-License-Identifier: LGPL-3.0-or-later
// Copyright (C) 2024 Sebouh Pual
```

veprbl replied:
Where did you copy this from?

sebouh137 commented:
> Thanks for the development, Sebouh! I have a minor fix suggested on the code itself, but I have some larger conceptual suggestions:
> 
> * I think this might make more sense alongside the `reco` algorithms since this is taking the final step from detector objects (topoclusters) to physics objects (neutron candidates)
Ok, I have moved it there.  
> * I would suggest renaming this to something like `FarForwardNeutronReconstrution` since neutrons will also be reconstructed elsewhere in ePIC with very different algorithms.
Ok, I have changed the name.

ruse-traveler commented on src/detectors/ZDC/ZDC.cc:
```diff
@@ -195,6 +195,13 @@ extern "C" {
           app   // TODO: Remove me once fixed
         ));
 
+        app->Add(new JOmniFactoryGeneratorT<FarForwardNeutronReconstruction_factory>(
+           "HcalFarForwardZDCNeutronCandidates",
+          {"HcalFarForwardZDCClusters"},  // edm4eic::ClusterCollection
+          {"HcalFarForwardZDCNeutronCandidates"}, // edm4eic::ReconstrutedParticleCollection,
+          app   // TODO: Remove me once fixed
+        ));
+
```
Whoops! Should've commented on this in my last review: alongside moving the algo/factory to reco, I would suggest moving this to the reco plugin (`global/reco/reco.cc`) and renaming the output to `ReconstructedFarForwardNeutrons` (or similar).

sebouh137 replied:
Done, thanks!

ruse-traveler replied:
Have you pushed the changes? I'm still seeing the factory declaration in `zdc.cc`...

sebouh137 replied:
oops; made the change locally but forgot to git add; git commit ..., before doing git pull; git push.   Pushed to the github now.

ruse-traveler replied:
Thanks! One final suggestion and then I'm happy to merge! I'd suggest renaming the output to `"ReconstructedFarForwardZDCNeutrons"` just to make it consistent with most of the other reconstructed particle collections

sebouh137 replied:
done

ruse-traveler reviewed (changes requested):
Nice! Thanks for incorporating the feedback! I see the conflicts and copyrights are fixed, so aside from the one suggestion I forgot to include in my last review, I'm happy with this!

sebouh137 commented on src/algorithms/calorimetry/NeutronReconstructionConfig.h:
```diff
@@ -0,0 +1,12 @@
+// Subject to the terms in the LICENSE file found in the top-level directory.
```
I think this was in one of the other Config.h files, but I forget which one.  However, since I have removed this file (due to there being no parameters), it is no longer relevant.
